### ISSUE TITLE
 memory for state

### ISSUE THREAD

**Issue by fi000** [keon/deep-q-learning#14]
thanks Keon for your great code!
I have two questions:
1- What does [0] means in self.model.predict(next_state)[0] and return np.argmin(act_values[0])? Does this mean that first element of batch?
2-If in addition to batch, I need that my state is the state from K times before, what is the necessary change in order to do this? I want to send the state=state[i-k+1]....state[i-1],state[i] not only one state! How I can do this?

Thanks again

**Comment by CarterEllsworth** [comment]
1. [Keras Model API](https://keras.io/models/model/) tell us that `model.predict(predict_batch)` returns a numpy array with an array of predictions for each element in `predict_batch`. Since `model.predict` is being called on a single array, `next_state` then we want the first and only element in the prediction array. Hence `'[0]'`

**Comment by fi000** [comment]
Thanks for question 1- I did not understand how to do it?
How to have a memory on state? in nature paper the memory is used and it is equal to 4! and this is in addition to using batch

**Comment by pskrunner14** [comment]
@fi000 can you provide a link to the said paper.

1. `self.model.predict(next_state)[0]` predicts on batch as @CarterEllsworth pointed out. It returns an array of predictions for each of the elements in the batch but since we're only predicting on one state element we only the first and only prediction, hence the `[0]`.

1. You could somehow normalize over the last `k` states as I've implemented [here](https://github.com/pskrunner14/trading-bot/blob/290cf9904d96777776b411b5a374f1b27d180d71/utils.py#L51). You'll need to adjust the dimensions according to whatever best suits your task.

**Comment by fi000** [comment]
@pskrunner14 You can refer to paper "Playing Atari with Deep Reinforcement Learning" section 4.1 last sentences of first paragraph! I have an state with 5 inputs but I have a problem in giving for instance 4 states in a frame! How we could do this in this code?

**Comment by keon** [comment]
1) Yes — `[0]` strips the leading batch dimension. The model takes inputs shaped `[batch, state_size]` and returns `[batch, action_size]`; `[0]` extracts the single sample. 2) For k-step state stacking, see #20 — keep a deque of the last k states and stack/concatenate them before passing to the network.
